Update for tensorflow 0.10.0rc0

mirosval opened the pull request:
Please if somebody could review this. 

For example I had to rename Cast op to CastOp and Const op to ConstOp otherwise it wouldn’t compile.

I’m pretty sure also other things need to be improved/fixed as I don’t really understand all of the ramifications of these changes

Mistobaan commented on tensorflow/src/main/java/org/bytedeco/javacpp/presets/tensorflow.java:
```diff
@@ -22,23 +22,19 @@
 import org.bytedeco.javacpp.tools.InfoMapper;
 
+import java.lang.annotation.*;
+
```
I this is preferable to have all the imports explicit.

Mistobaan commented on tensorflow/tensorflow-0.10.0rc0.patch:
```diff
@@ -0,0 +1,25 @@
+ )
+
++cc_binary(
++    name = "libtensorflow.so",
```
There is already a [target for the c library](https://github.com/tensorflow/tensorflow/blob/2a10e469d70827ede21c0bb5d0a46aac97404d4c/tensorflow/BUILD#L177). Why do we need to add a new one ?

mirosval replied:
I have no idea, I just copied what was there previously and made minimal changes needed in order to build it. I put up this PR so that we can fix these issues. So thank you for your comments, they are very appreciated, I'll look into it and see what I can improve :)

saudet replied:
They've added that target only recently. If it does what we need, let's use that, sure. In any case, remove the patch for v0.9.0. Thanks!

mirosval commented:
I applied the suggested changes, but it turns out that there are compilation problems at a later stage. I'm really not sure how to deal with them. 

```
javacpp-presets/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:85747:53: error: call to implicitly-deleted default constructor of
      '::tensorflow::CompositeOpScopes'
        ::tensorflow::CompositeOpScopes* rptr = new ::tensorflow::CompositeOpScopes();
                                                    ^
javacpp-presets/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0rc0/tensorflow/cc/framework/scope.h:253:9: note: default constructor of 'CompositeOpScopes'
      is implicitly deleted because field 'child' has no default constructor
  Scope child;
        ^
javacpp-presets/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:85774:16: error: object of type 'tensorflow::Scope' cannot be assigned because
      its copy assignment operator is implicitly deleted
    ptr->child = *ptr0;
               ^
javacpp-presets/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0rc0/tensorflow/cc/framework/scope.h:238:37: note: copy assignment operator of 'Scope' is
      implicitly deleted because field 'control_deps_' has no copy assignment operator
  const std::vector<ops::Operation> control_deps_;
                                    ^
javacpp-presets/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:85799:53: error: call to implicitly-deleted default constructor of
      '::tensorflow::CompositeOpScopes'
        ::tensorflow::CompositeOpScopes* rptr = new ::tensorflow::CompositeOpScopes[arg0];
                                                    ^
javacpp-presets/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0rc0/tensorflow/cc/framework/scope.h:253:9: note: default constructor of 'CompositeOpScopes'
      is implicitly deleted because field 'child' has no default constructor
  Scope child;
        ^
```

mirosval commented:
Another issue is that there seem to be two ops `Cast` and `Const` which obviously causes problems because of the naming clashes. 

I tried:
```
.put(new Info("Cast").skip())
// or
.put(new Info("Cast").pointerTypes("CastOp"))
```

but both finished with an error. It either can not be transformed to java in tensorflow.java, because `Cast` conflicts with an annotation name, or it will fail later when compiling the C++ JNI because I renamed it to CastOp and it can't find such an OP.

Roughly the same problem is with `Const`

saudet commented:
It works for me with the following, but it looks like a lot has changed since version 0.9.0. The C++ compiler complains about of bunch of unrelated errors that we'll need to fix.
```java
.put(new Info("tensorflow::ops::Cast").pointerTypes("CastOp"))
.put(new Info("tensorflow::ops::Const").pointerTypes("ConstOp"))
```

Also, building the "//tensorflow:libtensorflow.so" doesn't get us the C++ ops API, but "//tensorflow:libtensorflow_cc.so" does so we should probably be using that.

mirosval commented:
I have been able to compile it if I skip or remove all of the offending C++ types. I can live without the Ops, but I need the `Env` class which seems to depend on e.g. `RandomAccessFile`. The problem is that `RandomAccessFile` comes in as `std::unique_ptr<::tensorflow::RandomAccessFile>` and I have no idea how to convert it.

I have tried:
```
.put(new Info("std::unique_ptr<tensorflow::RandomAccessFile>").pointerTypes("RandomAccessFiles"))
```

but this results in:
```
javacpp-presets/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:70137:90: error: cannot initialize a parameter of type 'std::unique_ptr<RandomAccessFile> *' with an lvalue of type '::tensorflow::RandomAccessFile *'
        rptr = new ::tensorflow::Status(ptr->NewRandomAccessFile((std::string&)adapter0, ptr1));
                                                                                         ^~~~
javacpp-presets/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0rc0/tensorflow/core/platform/env.h:88:65: note: passing argument to parameter 'result' here
                             std::unique_ptr<RandomAccessFile>* result);
```

If I manually edit `jnitensorflow.cpp` and replace:
```
rptr = new ::tensorflow::Status(ptr-> NewRandomAccessFile((std::string&)adapter0, ptr1));
```
with:
```
rptr = new ::tensorflow::Status(ptr-> NewRandomAccessFile((std::string&)adapter0, (std::unique_ptr<::tensorflow::RandomAccessFile> *)ptr1));
```

then it compiles, but I have no idea either how to achieve this with the `Info` or if it is even the right thing to do.

How can I represent unique pointers?

saudet commented:
Just now, I added support for `unique_ptr`. Not tested with TensorFlow yet, but after defining `UNIQUE_PTR_NAMESPACE` to `std`, it should pick it up.

andreas-eberle commented:
What's the status on this?

mirosval commented:
Waiting for Tensorflow 0.10 to be released.

andreas-eberle commented:
But your current version (this PR) is fully working with the current RC?

mirosval commented:
It works for me™. However I had to remove some ops in order to get it to work. But if your goal is to open a session, load a model and then run predictions, then that has worked for me.

andreas-eberle commented:
Do you have a branch where you have removed those ops / a branch that is compiling?

mirosval commented:
@andreas-eberle I merged master and resolved the conflicts. Now it compiles for me with `mvn install --projects .,tensorflow`

juraj-bicikl commented:
Hi, I tried to compile it using "mvn install --projects .,tensorflow" on a Mac, however I got this (during linking stage):

ld: library not found for -ltensorflow
clang: error: linker command failed with exit code 1 (use -v to see invocation)

This step works fine on previous javacpp-presets commits (with tensorflow 0.9).

mirosval commented:
That has to be something trivial, like a botched tensorflow install.

mirosval commented:
@saudet I don't know why google is not tagging TF final release on GitHub, but on the website they have r0.10 as the default when installing. Also there have been no substantial commits in the past weeks to r0.10. So I have switched to the head of the r0.10 branch and built from there. It seems fairly stable now, so maybe you can merge this...

saudet commented:
Could we try to get the ops back? I'm sure someone somewhere needs them...

Mistobaan commented:
seems that r0.10 is still in alpha

mirosval commented:
@saudet  I'm trying to add the ops back, but having more problems:

CPP:
```
typedef std::unordered_map<string, std::pair<int, int>> NameRangeMap;
Status NameRangesForNode(const NodeDef& node_def, const OpDef& op_def,
                         NameRangeMap* inputs, NameRangeMap* outputs);
```

Java:
```
.put(new Info("tensorflow::NameRangeMap*", "std::unordered_map<std::string,std::pair<int,int> >").pointerTypes("NameRangeMap").define())
```

Generated into:
```
@Namespace("tensorflow") public static native @ByVal Status NameRangesForNode(@Const @ByRef NodeDef node_def, @Const @ByRef OpDef op_def,
                         NameRangeMap inputs, NameRangeMap outputs);
```

But here `NameRangeMap` does not work because it's not defined. Its the typedef from above. I have tried different variations but seems that it never gets defined in the generated code?

@Mistobaan based on what information? They released the RC more than a month ago and it seems to me that the development in the r0.10 has halted to a crawl in the past two weeks, while master is running fast ahead.

Mistobaan commented:
@mirosval I don't think they have announced the release and the one on github is still marked Pre-Release https://github.com/tensorflow/tensorflow/releases

Mistobaan commented:
https://github.com/tensorflow/tensorflow/releases/tag/v0.10.0
They just released it

On Monday, September 12, 2016, Miroslav Zoričák <notifications@github.com>
wrote:

> @saudet <https://github.com/saudet> I'm trying to add the ops back, but
> having more problems:
>
> CPP:
>
> typedef std::unordered_map<string, std::pair<int, int>> NameRangeMap;
> Status NameRangesForNode(const NodeDef& node_def, const OpDef& op_def,
>                          NameRangeMap* inputs, NameRangeMap* outputs);
>
> Java:
>
> .put(new Info("tensorflow::NameRangeMap*", "std::unordered_map<std::string,std::pair<int,int> >").pointerTypes("NameRangeMap").define())
>
> Generated into:
>
> @Namespace("tensorflow") public static native @ByVal Status NameRangesForNode(@Const @ByRef NodeDef node_def, @Const @ByRef OpDef op_def,
>                          NameRangeMap inputs, NameRangeMap outputs);
>
> But here NameRangeMap does not work because it's not defined. Its the
> typedef from above. I have tried different variations but seems that it
> never gets defined in the generated code?
>
> @Mistobaan <https://github.com/Mistobaan> based on what information? They
> released the RC more than a month ago and it seems to me that the
> development in the r0.10 has halted to a crawl in the past two weeks, while
> master is running fast ahead.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bytedeco/javacpp-presets/pull/266#issuecomment-246399276>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAG3147YRL2MOhbZXtmEl4iC88zNxPzJks5qpXlWgaJpZM4JiJ3J>
> .
>


-- 
LinkedIn: http://linkedin.com/in/fmilo
Twitter: @fabmilo
Github: http://github.com/Mistobaan/
-----------------------
Simplicity, consistency, and repetition - that's how you get through. (Jack
Welch)
Perfection must be reached by degrees; she requires the slow hand of time
(Voltaire)
The best way to predict the future is to invent it (Alan Kay)

saudet commented:
@mirosval We don't need to provide Info for typedefs, so this should work better:
```
.put(new Info("std::unordered_map<std::string,std::pair<int,int>>").pointerTypes("NameRangeMap").define())
```
Also `std::unordered_map` wasn't registered as a "basic container", so I've just added that (in the commit above). Thanks for looking into this!

mirosval commented:
@saudet  I keep having problems with the unordered_map, the snippet from your comment seems to generate:

Java
```
@Name("std::unordered_map<std::string,std::pair<int,int> >") public static class NameRangeMap extends Pointer {
    static { Loader.load(); }
    /** Pointer cast constructor. Invokes {@link Pointer#Pointer(Pointer)}. */
    public NameRangeMap(Pointer p) { super(p); }
    public NameRangeMap(int[] firstValue, int[] secondValue) { this(Math.min(firstValue.length, secondValue.length)); put(firstValue, secondValue); }
    public NameRangeMap()       { allocate();  }
    private native void allocate();
    public native @Name("operator=") @ByRef NameRangeMap put(@ByRef NameRangeMap x);

    public native long size();

    @Index public native int first(@StdString BytePointer i); public native NameRangeMap first(@StdString BytePointer i, int first);
    @Index public native int second(@StdString BytePointer i);  public native NameRangeMap second(@StdString BytePointer i, int second);

    public NameRangeMap put(int[] firstValue, int[] secondValue) {
        for (int i = 0; i < firstValue.length && i < secondValue.length; i++) {
            first(i, firstValue[i]);
            second(i, secondValue[i]);
        }
        return this;
    }
}
```

You'll notice that `int[], int[]` is obviously not the type we wanted there. 

Another issue is the spaces, I noticed that it makes a difference if the space is there at the end between `> >` or not.

mirosval commented:
And more problems, this is really frustrating:

Original CPP:
```
  InferenceContext(const NodeDef* node_def, const OpDef& op_def,
                   const std::vector<string>& input_shapes,
                   const std::vector<const Tensor*>& input_tensors);
```

Generated Java:
```
public InferenceContext(@Const NodeDef node_def, @Const @ByRef OpDef op_def,
                     @Const @ByRef StringVector input_shapes,
                     @Cast("const tensorflow::Tensor**") @StdVector PointerPointer input_tensors) { super((Pointer)null); allocate(node_def, op_def, input_shapes, input_tensors); }
```

Error:
```
error: no matching constructor for initialization of 'VectorAdapter<tensorflow::Tensor *>'
    VectorAdapter< tensorflow::Tensor* > adapter3((const tensorflow::Tensor**)ptr3, size3, owner3);
                                         ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
javacpp-presets/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:554:5: note: candidate constructor not viable: no known conversion from 'const tensorflow::Tensor **' to 'tensorflow::Tensor *const *' for 1st argument
```

It seems like this `const tensorflow::Tensor **` is incorrect, but I have no idea how to fix it.

mirosval commented:
@saudet I have managed to solve most of the issues and I re-added most of the ops. There are still some problems, such as apparently __COUNTER__ macro is not expanded properly. I added comments above each `skip()` that I added with errors that occur when I remove it.

I will see if I can clean it up a little bit more, but generally I think its good to go.

saudet commented:
I've made a few updates to JavaCPP, so this works better now:
```java
.put(new Info("std::unordered_map<std::string,std::pair<int,int> >").pointerTypes("NameRangeMap").define())
```
(Issues with spaces are known: #231)

For `std::vector` types of any complexity,`@StdVector` shouldn't be used. Defining a class like this should give better results:
```java
.put(new Info("std::vector<const Tensor*>").pointerTypes("TensorPointerVector").define())
```
For missing destructors, we can use the `@NoDeallocator` annotation as shown here:
https://github.com/bytedeco/javacpp/issues/124#issuecomment-239774524

`__COUNTER__` should probably be added to the `.cppTypes().annotations()` list at the top.

And "cannot find symbol" are probably caused by missing header files from the include list at the top.

Let me know if you encounter any issues with that or otherwise! Thanks for your time on this

mirosval commented:
Thank you for all your help!

The `__COUNTER__` shenanigans throws an error even after adding it to the top list, it would seem that the macro does not expand when its a part of an identifier. 

```
use of undeclared identifier 'register_op__COUNTER__'
```

Essentially all the remaining problems are a variation of the same: I don't know how to deal with typedeffed function pointers. I have tried adding a class extending the `FunctionPointer` for them, but then I just run into more problems later on.
Example:

Java:
```
.put(new Info("std::function<tensorflow::Status(tensorflow::shape_inference::InferenceContext*)>").pointerTypes("ShapeInferenceFn"))

public static class ShapeInferenceFn extends FunctionPointer {
        static { Loader.load(); }
        /** Pointer cast constructor. Invokes {@link Pointer#Pointer(Pointer)}. */
        public    ShapeInferenceFn(Pointer p) { super(p); }
        protected ShapeInferenceFn() { allocate(); }
        private native void allocate();
        public native org.bytedeco.javacpp.tensorflow.Status call(@Cast("shape_inference::InferenceContext*") Pointer node);
    }
```

Cpp:
```
typedef std::function<Status(shape_inference::InferenceContext* c)>
    OpShapeInferenceFn;

struct OpRegistrationData {
 public:
  OpRegistrationData() {}
  OpRegistrationData(const OpDef& def) : op_def(def) {}

  OpDef op_def;
  OpShapeInferenceFn shape_inference_fn;
};
```

jnitensorflow.cpp (last line)
```
::tensorflow::OpRegistrationData* ptr = (::tensorflow::OpRegistrationData*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
    if (ptr == NULL) {
        env->ThrowNew(JavaCPP_getClass(env, 17), "This pointer address is NULL.");
        return 0;
    }
    jlong position = env->GetLongField(obj, JavaCPP_positionFID);
    ptr += position;
    jobject rarg = NULL;
    JavaCPP_org_bytedeco_javacpp_presets_tensorflow_00024ShapeInferenceFn* rptr;
    rptr = new (std::nothrow) JavaCPP_org_bytedeco_javacpp_presets_tensorflow_00024ShapeInferenceFn;
    if (rptr != NULL) rptr->ptr = &ptr->shape_inference_fn;
```

Produces:
```
error: assigning to '::tensorflow::Status *(*)(shape_inference::InferenceContext *)' from incompatible type 'OpShapeInferenceFn *' (aka 'function<tensorflow::Status (shape_inference::InferenceContext *)> *')
    if (rptr != NULL) rptr->ptr = &ptr->shape_inference_fn;
                                ^ ~~~~~~~~~~~~~~~~~~~~~~~~
```

saudet commented:
When macros don't expand properly like that, it's usually because there there is some macros it depends on that are defined in other header files that are missing from the include list...

`FunctionPointer` is not a `std::function`, but an `std::function` can be constructed from a function pointer, so we need to force those to `@MemberSetter` only without "getters". Check the other examples that are already there and just do the same. :)

mirosval commented:
@saudet The `__COUNTER__` macro is supposed to be built in the preprocessor, I couldn't find it handled anywhere in `javacpp`, so I think it's just not handled properly. All the ops use this macro to "register" themselves and if I ignore it they just don't get registered.

```cpp
#define REGISTER_OP(name) REGISTER_OP_UNIQ_HELPER(__COUNTER__, name)
#define REGISTER_OP_UNIQ_HELPER(ctr, name) REGISTER_OP_UNIQ(ctr, name)
#define REGISTER_OP_UNIQ(ctr, name)                                          \
  static ::tensorflow::register_op::OpDefBuilderReceiver register_op##ctr    \
      TF_ATTRIBUTE_UNUSED =                                                  \
          ::tensorflow::register_op::OpDefBuilderWrapper<SHOULD_REGISTER_OP( \
              name)>(name)
```

I'm not entirely sure why they did it like this, what does the static register_op_1 do, it just seems to be there to anchor the OP so that it doesn't get immediately optimized away by the compiler? 

Maybe it would be enough to get rid of the `static ::tensorflow::register_op::OpDefBuilderReceiver register_op##ctr TF_ATTRIBUTE_UNUSED =` part? So for our purposes we just run the expanded registration?

Maybe it doesn't make sense to try to parse this at all and the ops can be retrieved from wherever they get registered?

saudet commented:
Oh, that's interesting. I didn't know about that:
https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html
So, I guess I'll be implementing that, but like you said maybe we don't need it.
Everyone, what's your opinion about that?

sonalgupta commented:
is it possible to merge this PR? I need some changes that are in the Tensorflow 0.10.0 release. thanks!

mirosval commented:
@saudet After some more testing it turns out that I have runtime problems since I re-introduced the ops. I'm pretty sure it's because something got messed up during generation, because at [6834f6](https://github.com/mirosval/javacpp-presets/commit/6834f6a3d383bc0ad7e30a550d82ee317e51bae3) it still works.

This seems very labor intensive, I think we will drop running predictions from Scala in favour of Python, the added maintenance and opaqueness is just not worth the (possibly) faster execution.

saudet commented:
I don't think the issue you're having would be any more transparent if it happened with the Python bindings...

In any case, thank you very much for your time on this! And if you could provide failing sample code, that would be great.

mirosval commented:
Using Python both for training and prediction will save me from maintaining an additional set of bindings that I don't fully understand.

In my prediction code I just create a session, load a graph from a binary protobuf file, feed an array of floats into the graph and collect an array of floats at the end of the graph. But somehow by adding those ops I get an error about "Duplicate gradient in Shape" `core/framework/function.cc:885`. 

I will try looking into this in more detail and hopefully we can somehow finish this PR together.

saudet commented:
Thanks :) Didn't you say you had runtime problems as in it compiles but doesn't run with a given workload?

I'll start by adding support for __COUNTER__ ...

mirosval commented:
Perhaps I wasn't clear enough, the "Duplicate gradient in Shape" error occurs when I try to run the (successfully built) tensorflow.jar using my model in the context of my service. The javacpp-presets build is successful, but then it fails when running predictions.

The same Scala code I use to run the TF model works when I build javacpp-presets at [6834f6](https://github.com/mirosval/javacpp-presets/commit/6834f6a3d383bc0ad7e30a550d82ee317e51bae3) and fails when I run e.g. the current HEAD of this PR. The main thing that changed between these two was the addition of the TF ops.

sonalgupta commented:
I was able to build @mirosval's code with javacpp 1.2.4. But got the exact same error of `F tensorflow/core/framework/function.cc:885] Check failed: GetOpGradFactory()->insert({op, func}).second Duplicated gradient for Shape` I trained my RNN model in Python and am trying to load the graph in Java.

mirosval commented:
It should work if you check out at the above mentioned commit. Except you won't have access to the ops.

saudet commented:
Ah, here's the issue. We need to build the "libtensorflow_cc.so" target, not "libtensorflow.so". The patch in the presets for old versions of TensorFlow included the ops in "libtensorflow.so", but they've since made them available as part of "libtensorflow_cc.so" instead. By building "libtensorflow_cc.so" we also get all the header files for the ops. We don't need to include, parse, and compile the `*.cc` source files. The build process is supposed to do that for us and generate simple `*.h` header files.

Similarly, the `REGISTER_GRADIENT_OP` etc macros only get used in `math_grad.cc` and they are not exported (unused) so we can safely ignore that.

andreas-eberle commented:
What is the status on this? @mirosval: Does this version fully work for you now? If yes, is there something else preventing this from being merged?

mirosval commented:
It does work for me. However I only use a subset of the functionality (create session, load protobuf, run prediction), so take it with a grain of salt. 

I think generally it could be merged, but that's @saudet's call.

andreas-eberle commented:
I tried to build this PR and had some issues:

* When I try to execute `mvn install --projects .,tensorflow -Djavacpp.cppbuild.skip=true`, I get the following error for every other project than tensorflow
```
[ERROR]   The project org.bytedeco.javacpp-presets:cuda:8.0-${project.parent.version} (/Users/aeberle/programming/javacpp-presets/cuda/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.bytedeco.javacpp-presets:cuda:8.0-${project.parent.version}: Failure to find org.bytedeco:javacpp-presets:pom:1.2.2 in http://maven.repository.redhat.com/earlyaccess/all/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-earlyaccess-repository has elapsed or updates are forced and 'parent.relativePath' points at wrong local POM @ line 6, column 11 -> [Help 2]
```
I was able to fix this by updating the version of the parent pom in all the other poms to the same one used by the tensorflow module. Is there another way than setting all these versions manually?

* The javacpp version used in this PR does not include the latest changes regarding the __COUNTER__ stuff mentioned by @saudet. Only the 1.2.5-SNAPSHOT currently contains these changes. After I built that version manually and adapted all the versions in the poms, I now face the next problem.

* When executing `mvn install --projects .,tensorflow -Djavacpp.cppbuild.skip=true` after applying the changes mentioned above, I get the following error
```
In file included from /Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:107:
In file included from /Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0/bazel-genfiles/tensorflow/core/lib/core/error_codes.pb.h:22:
In file included from /Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0/bazel-tensorflow-0.10.0/external/protobuf/src/google/protobuf/arena.h:52:
In file included from /Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0/bazel-tensorflow-0.10.0/external/protobuf/src/google/protobuf/stubs/atomic_sequence_num.h:33:
In file included from /Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0/bazel-tensorflow-0.10.0/external/protobuf/src/google/protobuf/stubs/atomicops.h:199:
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/cppbuild/macosx-x86_64/tensorflow-0.10.0/bazel-tensorflow-0.10.0/external/protobuf/src/google/protobuf/stubs/atomicops_internals_macosx.h:173:9: warning: 
      'OSAtomicCompareAndSwap64Barrier' is deprecated: first deprecated in macOS 10.12 - Use std::atomic_compare_exchange_strong() from <atomic> instead [-Wdeprecated-declarations]
    if (OSAtomicCompareAndSwap64Barrier(
        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/libkern/OSAtomicDeprecated.h:645:9: note: 
      'OSAtomicCompareAndSwap64Barrier' has been explicitly marked deprecated here
bool    OSAtomicCompareAndSwap64Barrier( int64_t __oldValue, int64_t __newValue,
        ^
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:933:72: warning: cannot delete expression with
      pointer-to-'void' type 'void *' [-Wdelete-incomplete]
static void JavaCPP_org_bytedeco_javacpp_Pointer_deallocate(void *p) { delete (void*)p; }
                                                                       ^      ~~~~~~~~
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:2591:24: warning: 
      'JavaCPP_org_bytedeco_javacpp_presets_tensorflow_00024EdgeLabelFunction_allocate_callback' has C-linkage specified, but returns user-defined type 'std::string &' (aka
      'basic_string<char, char_traits<char>, allocator<char> > &') which is incompatible with C [-Wreturn-type-c-linkage]
JNIEXPORT std::string& JavaCPP_org_bytedeco_javacpp_presets_tensorflow_00024EdgeLabelFunction_allocate_callback(const tensorflow::Edge* arg0) {
                       ^
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:2714:24: warning: 
      'JavaCPP_org_bytedeco_javacpp_presets_tensorflow_00024NodeLabelFunction_allocate_callback' has C-linkage specified, but returns user-defined type 'std::string &' (aka
      'basic_string<char, char_traits<char>, allocator<char> > &') which is incompatible with C [-Wreturn-type-c-linkage]
JNIEXPORT std::string& JavaCPP_org_bytedeco_javacpp_presets_tensorflow_00024NodeLabelFunction_allocate_callback(const tensorflow::Node* arg0) {
                       ^
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:14119:32: warning: 
      'JavaCPP_org_bytedeco_javacpp_tensorflow_00024FalseOpDefBuilderWrapper_00024Fn_1InferenceContext_allocate_callback' has C-linkage specified, but returns user-defined type
      '::tensorflow::Status' which is incompatible with C [-Wreturn-type-c-linkage]
JNIEXPORT ::tensorflow::Status JavaCPP_org_bytedeco_javacpp_tensorflow_00024FalseOpDefBuilderWrapper_00024Fn_1InferenceContext_allocate_callback(::tensorflow::shape_inference::Inf...
                               ^
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:14831:32: warning: 
      'JavaCPP_org_bytedeco_javacpp_tensorflow_00024TrueOpDefBuilderWrapper_00024Fn_1InferenceContext_allocate_callback' has C-linkage specified, but returns user-defined type
      '::tensorflow::Status' which is incompatible with C [-Wreturn-type-c-linkage]
JNIEXPORT ::tensorflow::Status JavaCPP_org_bytedeco_javacpp_tensorflow_00024TrueOpDefBuilderWrapper_00024Fn_1InferenceContext_allocate_callback(::tensorflow::shape_inference::Infe...
                               ^
/Users/aeberle/Documents/programming/java/dekra/javacpp-presets-clean/tensorflow/target/classes/org/bytedeco/javacpp/jnitensorflow.cpp:16425:32: warning: 
      'JavaCPP_org_bytedeco_javacpp_tensorflow_00024OpDefBuilder_00024Fn_1InferenceContext_allocate_callback' has C-linkage specified, but returns user-defined type '::tensorflow::Status'
      which is incompatible with C [-Wreturn-type-c-linkage]
JNIEXPORT ::tensorflow::Status JavaCPP_org_bytedeco_javacpp_tensorflow_00024OpDefBuilder_00024Fn_1InferenceContext_allocate_callback(::tensorflow::shape_inference::InferenceContex...
                               ^
17 warnings generated.
ld: library not found for -ltensorflow
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Do you have any idea what I'm doing wrong? Is it working for you with a clean setup?

mirosval commented:
I think you need to do `mvn install --projects .,tensorflow`, not `mvn install --projects .,tensorflow -Djavacpp.cppbuild.skip=true`. You skipped the build, so your tensorflow lib was not built.

andreas-eberle commented:
@mirosval: Sorry, that was a copy and paste failure. I actually ran it without the skip option (I updated my question). I just also tried to first manually run the cppbuild.sh script and skip that step in the maven process. Any other ideas?

mirosval commented:
I run both, `cppbuild.sh` and after that `mvn install...`. 

Clearly it looks like your tensorflow was not built, or it can not be found in that path. 

I did come across a similar problem on Linux, where I had to change the path where bazel put header files.

The best I can do for you is share [this gist](https://gist.github.com/mirosval/4413439c8dc42bcd645cfefa3b8d0da7) that contains the Dockerfile I used to build the linux version of the project.

andreas-eberle commented on tensorflow/src/main/java/org/bytedeco/javacpp/presets/tensorflow.java:
```diff
@@ -41,40 +35,91 @@
+        "tensorflow/cc/framework/ops.h",
+        "tensorflow/cc/framework/cc_op_gen.h",
+        "tensorflow_adapters.h"},
+        link = "tensorflow"),
```
@mirosval: I think I found it. The build worked when I changed to link agains `tensorflow_cc`. I guess this is because you had to switch to build libtensorflow_cc instead of libtensorflow which it was in version 0.9.

saudet commented:
We need to change the versions in the pom.xml files, yes, sorry about that. I need to figure out a better way of dealing with this. I think I'll just start using a major version number for all pom.xml files on all master branches (like 1.3-SNAPSHOT) and when I need to do some minor release of a past version (say 1.2.x), I'll just do it as a branch.

Anyway, the build also works for me, but there are still some things to fix...

The ops header files are missing, at least these:
```
"tensorflow/cc/ops/const_op.h", "tensorflow/cc/ops/cc_op_gen.h", "tensorflow/cc/ops/array_ops.h",
"tensorflow/cc/ops/data_flow_ops.h", "tensorflow/cc/ops/image_ops.h", "tensorflow/cc/ops/io_ops.h",
"tensorflow/cc/ops/linalg_ops.h", "tensorflow/cc/ops/logging_ops.h", "tensorflow/cc/ops/math_ops.h",
"tensorflow/cc/ops/nn_ops.h", "tensorflow/cc/ops/parsing_ops.h", "tensorflow/cc/ops/random_ops.h",
"tensorflow/cc/ops/sparse_ops.h", "tensorflow/cc/ops/state_ops.h", "tensorflow/cc/ops/string_ops.h",
"tensorflow/cc/ops/training_ops.h", "tensorflow/cc/ops/user_ops.h"
```
And we probably don't need this:
```java
.put(new Info("const_op.h").skip())
.put(new Info("math_ops.h").skip())
.put(new Info("array_ops.h").skip())
```

This doesn't appear to be used:
```
.put(new Info("tensorflow::Scope").pointerTypes("Scope"))
.put(new Info("std::vector<tensorflow::ops::Input>").pointerTypes("InputVector"))
.put(new Info("std::vector<tensorflow::ops::Input>::iterator").skip())
.put(new Info("std::vector<tensorflow::ops::Input>::const_iterator").skip())
```

This is normal behavior for JavaCPP:
```java
// For some reason the #define this was in got parsed even though its #ifdef guard failed
// see "tensorflow/core/framework/selective_registration.h" for more details
.put(new Info("SHOULD_REGISTER_OP_GRADIENT").skip())
```
It will try to parse everything by default, because it usually works better that way. This is the way to fix this:
```java
.put(new Info("SELECTIVE_REGISTRATION").define(false))
```

I'm afraid this will also skip the `shape_inference_fn` field:
```java
.put(new Info("tensorflow::OpShapeInferenceFn").skip())
```
Something like this would be better:
```java
.put(new Info("tensorflow::OpRegistrationData::shape_inference_fn")
        .javaText("@MemberSetter public native OpRegistrationData shape_inference_fn(@ByVal ShapeInferenceFn shape_inference_fn);"))
```

The rest seems good :) Thanks!

mirosval commented:
Seems that the ops are back and it compiled for me. So assuming you update the javacpp version in all your poms to 1.2.5-SNAPSHOT it should compile for you as well.

However all your other points, like the skipped headers and iterators, etc don't seem to work for me, I always run into issues trying to remove them.

saudet commented:
Looks like `const_op.h` contains functions named `Const` that conflict with the new `Const` class, but they don't seem to be used anywhere anymore, so commenting that out solves this issue. I've patched a few other things: [tensorflow_20160929_diff.txt](https://github.com/bytedeco/javacpp-presets/files/499661/tensorflow_20160929_diff.txt)
Looks good? Seems to work well enough like that for me.

mirosval commented:
Everything works after including your patch, so I think from my side we're good to go. 

I have allowed edits from maintainers, so in case there are some other changes necessary you should be able to push directly into my master branch.

Mistobaan commented:
The Const, Cast and few other static import operation does not work anymore. The version of the tensorflow in the Readme should be updated to 0.10.0. Somewhere we have to enforce that the javacpp version MUST be 1.2.5 or greater.

saudet commented:
@mirosval Actually the Const functions are used in the `example_trainer.cc` sample, so we need to put those back in. I've updated JavaCPP for that, and with these info, we can work around the name clashes: 
```java
.put(new Info("tensorflow::ops::Cast").cppTypes("class tensorflow::ops::Cast").pointerTypes("CastOp"))
.put(new Info("tensorflow::ops::Const").cppTypes("class tensorflow::ops::Const").pointerTypes("ConstOp"))
```

I didn't notice I had push access to your repo. Anyway, I'd like to get all those commits squashed, and it might as well be under your name. The following commands seem to work, but feel free to do it your way! Thanks
```
git reset 9947d3c861c98007acd67918f3206ad940de1bf3
git commit -a -m "Update for tensorflow 0.10.0"
git remote add upstream https://github.com/bytedeco/javacpp-presets/
git fetch upstream
git rebase upstream/master
git push --force
```

@Mistobaan Once this and a couple of other things are in, I'll be switching all versions to 1.3-SNAPSHOT, and of course update the README.md file after the next release, whenever that happens. Will also be working on the new mapping for Const and Cast later, but help would be welcome.

Mistobaan commented:
@saudet if you point me in the right direction on the work that needs to be done for Const & Cast I might be able to take a look at it.

mirosval commented:
@saudet I think you can set up the squashing directly on GitHub (see https://github.com/blog/2141-squash-your-commits)

I'm trying to put the Cast & Const back using your snippet, but I have run into an issue with DataType not being defined, despite 800 lines below in another method it gets correctly recognized and cast to int (it's an enum from core/framework/types.pb.h). It seems that with `ConstOp` and `CastOp` it doesn't get expanded for some reason from: `@ByVal DataType dtype` to `@Cast("tensorflow::DataType") int dtype` and then I get compilation error `tensorflow.java:[14295,85] cannot find symbol
  symbol:   class DataType
  location: class org.bytedeco.javacpp.tensorflow.ConstOp`

I'm on vacation next week so I won't have time to look into this.

saudet commented:
@mirosval Oh, I didn't realize GitHub had a squash option there now. Good to know. :) Let's try it out.
That error just means you forgot to install the latest SNAPSHOT of JavaCPP. It works fine.

@Mistobaan Thanks, will look into updating `ExampleTrainer`, see what needs to be done for that and we'll see.

saudet commented:
@Mistobaan Done: https://github.com/bytedeco/javacpp-presets/commit/dd4ffa747a2e6848271efd9f98ac17ec72597a94 The ops API got a bit uglier, so we might want to have create nice wrappers for that, but anyway it works. Let me know if there is anything missing!

Mistobaan commented:
getting this error now:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project tensorflow: Compilation failure: Compilation failure:
[ERROR] /home/fmilo/workspace/javacpp-presets/tensorflow/src/main/java/org/bytedeco/javacpp/tensorflow.java:[14391,85] cannot find symbol
[ERROR] symbol:   class DataType
[ERROR] location: class org.bytedeco.javacpp.tensorflow.ConstOp
[ERROR] /home/fmilo/workspace/javacpp-presets/tensorflow/src/main/java/org/bytedeco/javacpp/tensorflow.java:[14392,99] cannot find symbol
[ERROR] symbol:   class DataType
[ERROR] location: class org.bytedeco.javacpp.tensorflow.ConstOp
[ERROR] /home/fmilo/workspace/javacpp-presets/tensorflow/src/main/java/org/bytedeco/javacpp/tensorflow.java:[24354,67] cannot find symbol
[ERROR] symbol:   class DataType
[ERROR] location: class org.bytedeco.javacpp.tensorflow.CastOp
[ERROR] /home/fmilo/workspace/javacpp-presets/tensorflow/src/main/java/org/bytedeco/javacpp/tensorflow.java:[24355,82] cannot find symbol
[ERROR] symbol:   class DataType
[ERROR] location: class org.bytedeco.javacpp.tensorflow.CastOp
[ERROR] -> [Help 1]
```

saudet commented:
@Mistobaan That just means you have an old version of JavaCPP somewhere. Please try again with the latest version on the master branch.
